src/target/riscv/batch.c

```diff
@@ -13,58 +13,62 @@
 #define DMI_SCAN_MAX_BIT_LENGTH (DTM_DMI_MAX_ADDRESS_LENGTH + DTM_DMI_DATA_LENGTH + DTM_DMI_OP_LENGTH)
 #define DMI_SCAN_BUF_SIZE (DIV_ROUND_UP(DMI_SCAN_MAX_BIT_LENGTH, 8))
 
+#define BATCH_RESERVED_SCANS 4
+
 static void dump_field(int idle, const struct scan_field *field);
 
 struct riscv_batch *riscv_batch_alloc(struct target *target, size_t scans, size_t idle)
 {
-	scans += 4;
+	scans += BATCH_RESERVED_SCANS;
 	struct riscv_batch *out = calloc(1, sizeof(*out));
-	if (!out)
-		goto error0;
+	if (!out) {
+		LOG_ERROR("Failed to allocate struct riscv_batch");
+		return NULL;
+	}
+
 	out->target = target;
 	out->allocated_scans = scans;
 	out->idle_count = idle;
-	out->data_out = malloc(sizeof(*out->data_out) * (scans) * DMI_SCAN_BUF_SIZE);
+	out->last_scan = RISCV_SCAN_TYPE_INVALID;
+
+	out->data_out = NULL;
+	out->data_in = NULL;
+	out->fields = NULL;
+	out->bscan_ctxt = NULL;
+	out->read_keys = NULL;
+
+	out->data_out = malloc(sizeof(*out->data_out) * scans * DMI_SCAN_BUF_SIZE);
```

Comment:
I'm not sure about using `DMI_SCAN_BUF_SIZE` here. IMHO, using `DIV_ROUND_UP(riscv_get_dmi_scan_length(target), 8)` would result in a significant reduction in memory usage, since `abits` can be as low as 6, but it is assumed to be 64 here.

Reply:
Good catch, I agree. There is potential for saving some memory.

If you don't mind, I'd prefer to make this change separately. So for now I have added at least a `FIXME` comment next to the allocation to make it clear that there is a potential for improvement.

Reply:
IMHO, that's even better, since the change in question is out of scope of the commit.